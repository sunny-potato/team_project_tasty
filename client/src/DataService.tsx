import axios from 'axios';
import React, { useEffect, useState } from 'react';
axios.defaults.baseURL = 'http://localhost:3000/api/v2';

export type RecipeInfo = {
  id: number;
  name: string;
  meal_type: string;
  new: boolean;
  popular: boolean;
  description: string;
};

export type Ingredient = {
  ingredients_id: number;
  ingredient: string;
  amount: number;
  unit_id: number;
  unit: string;
};

export type Recipe = {
  recipeInfo: RecipeInfo;
  ingredients: Ingredient[];
};
export type Unit = {
  id: number;
  unit: string;
};
export type UnitsList = {
  units: Unit[];
};

export type EachIngredient = {
  id: number;
  ingredient: string;
};

export type IngredientsList = {
  ingredients: EachIngredient[];
};

class DataService {
  /* Get a specific recipe with known id */

  get(id: number | string) {
    // Use this for testing locally only
    // return new Promise<Recipe>((resolve, reject) => {
    //   resolve({
    //     recipeInfo: {
    //       id: 1,
    //       name: 'The best pasta in the world',
    //       meal_type: 'Vegan',
    //       new: true,
    //       popular: true,
    //       description:
    //         'Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolore, corporis. Distinctio deserunt aperiam hic voluptatum, ut a molestias cumque tenetur consequatur, ad dolores voluptatem provident eveniet, porro magnam est veritatis.',
    //     },
    //     ingredients: [
    //       { name: 'Olive oil', amount: 3, unit: 'tbsp' },
    //       { name: 'Onion', amount: 1, unit: 'pcs' },
    //       { name: 'Garlic', amount: 3, unit: 'cloves' },
    //       { name: 'Chopped Tomato', amount: 400, unit: 'g' },
    //       { name: 'Dried spaghetti', amount: 300, unit: 'g' },
    //     ],
    //   });
    // });

    return axios
      .get<Recipe>('/recipe/' + id)
      .then((response) => response.data)
      .catch((error) => {
        console.error(error);
        throw error;
      });
  }
  // Get an array of all recipes

  getAll() {
    return axios
      .get<[]>('/recipe/')
      .then((response) => response.data)
      .catch((error) => {
        console.error(error);
        throw error;
      });
  }
  /* Post new recepie */
  create(data: Recipe) {
    return axios
      .post<{ id: number }>('/recipe', { data })
      .then((response) => response.data.id)
      .catch((error) => {
        console.error(error);
        throw error;
      });
  }
  /* Edit an existing recipe by known id */

  edit(id: number) {
    return alert('function not implementet yet');

    /* Delete a recipe known id */
  }
  delete(id: number) {
    return axios
      .delete('/recipe/' + id)
      .then((response) => response.data)
      .catch((error) => {
        console.error(error);
        throw error;
      });
  }
  /* Get all ingredients */
  getAllIngredients() {
    return axios
      .get('/ingredient')
      .then((response) => response.data)
      .catch((error) => {
        console.error(error);
        throw error;
      });
  }

  /* Get all units */
  getAllUnits() {
    return axios
      .get<[]>('/unit')
      .then((response) => response.data)
      .catch((error) => {
        console.error(error);
        throw error;
      });
  }
}

const dataService = new DataService();
export default dataService;
