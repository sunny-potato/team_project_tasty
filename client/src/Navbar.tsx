import React from 'react';
import { Link } from 'react-router-dom';
import './index.css';

export function Navbar() {
  return (
    <nav className="NavMain">
      <Link to="/" className="Logo_title">
        Tasty
      </Link>
      <ul className="NavList">
        <li>
          <Link to="/" className="Link">
            Home
          </Link>
        </li>
        <li>
          <Link to="explore" className="Link">
            Explore
          </Link>
        </li>
        <li>
          <Link to="new" className="Link">
            Create new
          </Link>
        </li>
        <li>
          <Link to="cart" className="Link">
            Shopping list
          </Link>
        </li>
      </ul>
    </nav>
  );
}
