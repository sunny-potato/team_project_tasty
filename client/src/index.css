@import url('https://fonts.googleapis.com/css2?family=Prompt:wght@200;600&display=swap');

* {
  box-sizing: border-box;
}

body {
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu',
    'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  margin: 0;
  padding: 0;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New', monospace;
}

.NavMain {
  background-color: rgb(41, 211, 118);
  color: black;
  display: flex;
  justify-content: space-between;
  align-items: stretch;
  gap: 3rem;

  padding: 0 2rem;
}
.NavList {
  padding: 0;
  margin: 0;
  list-style: none;
  display: flex;
  gap: 2rem;
  text-decoration: none;
}
.NavList li:active {
  background-color: rgb(241, 250, 245);
}

.NavList li:hover {
  background-color: rgb(3, 190, 88);
}

.Link {
  font-family: 'Prompt', sans-serif;
  font-weight: 200;
  height: 100%;
  text-decoration: none;
  display: flex;
  align-items: center;
  padding: 0.5rem;
}

.Logo_title {
  font-family: 'Prompt', sans-serif;
  font-weight: 600;
  text-decoration: none;
  font-size: 2.5rem;
}
