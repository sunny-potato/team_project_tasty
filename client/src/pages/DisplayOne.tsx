import './PageStyling.css';
import React, { useEffect, useState, useRef, Component } from 'react';
import dataService, { Recipe, Ingredient, RecipeInfo } from '../DataService';
import { Link, useParams, useLocation } from 'react-router-dom';

export function DisplayOne() {
  const [recipe, setRecipe] = useState<Recipe>();
  const [ingredients, setIngredients] = useState<Ingredient[]>();

  //To get navigation by page and id
  const params = useParams();
  const id: string = params.id;

  // Get inital value in load, uses recipe id as source
  // Also loads values recipe and ingredients by setState

  useEffect(() => {
    let id: string = params.id;

    dataService.get(id).then((data) => {
      setRecipe(data);
      setIngredients(data.ingredients);
    });
  }, []);

  // Updates ingredients according to value chosen by user

  function changePortions(portions: number) {
    let newAmounts: Ingredient[];
    let defaultAmounts: Ingredient[] = recipe!.ingredients;

    newAmounts = defaultAmounts?.map((i) => ({
      ...i,
      amount: parseFloat(((i.amount / 4) * portions).toPrecision(2)),
    }));
    setIngredients(newAmounts);
  }

  return (
    <div className="Content-main">
      <h1 className="Recipe-title">{recipe?.recipeInfo.name}</h1>
      <h5 className="Recipe-tags">Tags: {recipe?.recipeInfo.meal_type}</h5>
      <h6 className="Recipe-new">{recipe?.recipeInfo.new ? 'This is a new recipe!' : null}</h6>
      <h6 className="Recipe-popular">
        {recipe?.recipeInfo.popular ? 'This item is popular right now!' : null}
      </h6>
      <p className="Recipe-description>">{recipe?.recipeInfo.description}</p>
      <div>
        <div>
          <label>Number of portions:&nbsp;</label>
          <select
            name="portions"
            defaultValue={4}
            onChange={(e) => changePortions(parseInt(e.currentTarget.value))}
          >
            <option value={1}>1</option>
            <option value={2}>2</option>
            <option value={3}>3</option>
            <option value={4}>4</option>
            <option value={5}>5</option>
            <option value={6}>6</option>
            <option value={7}>7</option>
            <option value={8}>8</option>
            <option value={9}>9</option>
            <option value={10}>10</option>
          </select>
        </div>
        <table className="Ingredients-main">
          <tbody>
            <tr>
              <th className="Ingredients-header">Ingredients:</th>
              <th></th>
            </tr>
            {ingredients?.map((content) => (
              <tr className="Ingredients-row" key={content.ingredients_id}>
                <td className="Ingredients-cell">
                  {content.amount}&nbsp;{content.unit}
                </td>

                <td className="Ingredients-name">{content.ingredient}</td>
              </tr>
            ))}
          </tbody>
        </table>
        <Link to="/">
          <button className="Button-navigation">Back</button>
        </Link>
        <Link to="/edit">
          <button className="Button-navigation">Edit recipe</button>
        </Link>
        <button
          className="Button-navigation"
          onClick={() => alert('Here will the like function be, not implemented yet')}
        >
          Like
        </button>
      </div>
    </div>
  );
}
