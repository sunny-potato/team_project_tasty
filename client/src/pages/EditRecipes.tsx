import React from 'react';
import { useParams } from 'react-router-dom';

export function EditRecipes() {
  return (
    <div>
      <h1>Site to edit the recipes</h1>
    </div>
  );
}
