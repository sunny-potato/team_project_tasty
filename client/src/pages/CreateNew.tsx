import React, { useState } from 'react';
import { RecipeInfo, Ingredient } from '../DataService';
import InputRecipeInfo from '../components/InputRecipeInfo';
import InputIngredients from '../components/InputIngredients';
import { useNavigate } from 'react-router-dom';
import { ValidateData, SaveEditedData } from '../components/SaveInputData';

const initialRecipeInfo: RecipeInfo = {
  id: undefined!,
  name: '',
  meal_type: '',
  new: false,
  popular: false,
  description: '',
};

const initialIngredient: Ingredient = {
  ingredients_id: undefined!,
  ingredient: '',
  amount: null,
  unit_id: 0,
  unit: 'initalUnit',
};

export function CreateNew() {
  const navigate = useNavigate();
  const [recipeInfo, setRecipeInfo] = useState<RecipeInfo>(initialRecipeInfo);
  const [ingredients, setIngredients] = useState<Ingredient[]>([initialIngredient]);
  const [isLoading, setIsLoading] = useState<boolean>(false);

  // console.log(recipeInfo);
  // console.log(ingredients);
  const onSubmit = (event: React.FormEvent) => {
    event.preventDefault();

    if (!isLoading) {
      setIsLoading(true);
      const isValid = ValidateData({ ingredients }); // current condition is always true
      if (isValid) {
        SaveEditedData({ ingredients, recipeInfo });
      }
    } else {
      alert('something wrong');
      //*********************//
      //here will be added f meal_type or description is required
    }
    setIsLoading(false);
  };

  return (
    <div>
      <h1 className="Page-title">Create new recipe</h1>
      <div className="Content-main">
        <form onSubmit={onSubmit}>
          <InputRecipeInfo recipeInfo={recipeInfo} setRecipeInfo={setRecipeInfo} />
          <InputIngredients ingredients={ingredients} setIngredients={setIngredients} />
          <div className="btn-group">
            <button type="submit" style={{ opacity: isLoading ? 0.3 : 1 }}>
              {isLoading ? 'Saving' : 'Save'}
            </button>
            <button type="button" className="cancelButton" onClick={() => navigate('/')}>
              Cancel
            </button>
          </div>
        </form>
      </div>
    </div>
  );
}
