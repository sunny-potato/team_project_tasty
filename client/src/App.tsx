import React from 'react';
import { Route, Routes } from 'react-router-dom';
import { CreateNew } from './pages/CreateNew';
import { DisplayOne } from './pages/DisplayOne';
import { EditRecipes } from './pages/EditRecipes';
import { Home } from './pages/Home';
import { Navbar } from './components/Navbar';
import './css/index.css';
import { Explore } from './pages/Explore';
import { ShoppingCart } from './pages/ShoppingCart';

function App() {
  return (
    <div>
      <Navbar />
      {/* @ts-ignore*/}
      <Routes basename="/">
        <Route path="/" element={<Home />} />
        <Route path="/recipe/:id" element={<DisplayOne />} />
        <Route path="/edit/:id" element={<EditRecipes />} />
        <Route path="/new" element={<CreateNew />} />
        <Route path="/explore" element={<Explore />} />
        <Route path="/cart" element={<ShoppingCart />} />
      </Routes>
    </div>
  );
}

export default App;
