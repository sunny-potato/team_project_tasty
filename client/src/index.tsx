import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
import App from './App';
import { HashRouter, BrowserRouter } from 'react-router-dom';

ReactDOM.render(
  /* Would prefer to change to server solution if avail instead of hashrouter */
  <HashRouter>
    <App />
  </HashRouter>,
  document.getElementById('root')
);
