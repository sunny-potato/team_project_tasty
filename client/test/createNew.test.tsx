import * as React from 'react';
import { shallow } from 'enzyme';
import { CreateNew } from '../src/pages/CreateNew';
import { MemoryRouter, Routes, Route } from 'react-router-dom';

import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { fireEvent } from '@testing-library/react';

import { Ingredient } from '../src/DataService';

let container: any = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('CreateNew page tests', () => {
  test('Create new, page draws correctly (snapshot)', async () => {
    await act(async () => {
      await render(
        <MemoryRouter>
          <CreateNew />
        </MemoryRouter>,
        container
      );
    });
    expect(container).toMatchSnapshot();
  });

  test('Input recipe info', async () => {
    await act(async () => {
      await render(
        <MemoryRouter>
          <CreateNew />
        </MemoryRouter>,
        container
      );
    });

    // set input name, and check if value gets updated
    let name = container.querySelector("input[name='name']");
    expect(name.value).toEqual('');
    fireEvent.change(name, {target: {value: 'New recipe'}});
    expect(name.value).toEqual('New recipe');

    // select meal type 'dinner' (first button) and check if active tag gets updated and background color changes
    let dinnerButton = container.querySelector("button");
    expect(dinnerButton.style.backgroundColor).toEqual('white');
    fireEvent.click(dinnerButton);
    expect(dinnerButton.style.backgroundColor).toEqual('lightblue');

    // write desctiption, and check that value gets updated
    let desctiption = container.querySelector("textarea[name='description']");
    expect(desctiption.value).toEqual('');
    fireEvent.change(desctiption, {target: {value: 'New description'}});
    expect(desctiption.value).toEqual('New description');

  });

  test('Input ingredients', async () => {
    
  })
});
