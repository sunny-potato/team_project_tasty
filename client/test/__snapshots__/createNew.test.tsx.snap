// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNew page tests Create new, page draws correctly (snapshot) 1`] = `
<div>
  <div>
    <h1
      class="Page-title"
    >
      Create new recipe
    </h1>
    <div
      class="Content-main"
    >
      <form>
        <div
          class="Content-recipeinfo"
        >
          <label
            class="recipe-name"
          >
             Name : 
          </label>
          <input
            autocomplete="off"
            name="name"
            required=""
            type="text"
            value=""
          />
          <div
            class="recipe-tag"
          >
            <div>
              <span>
                Tag :
              </span>
              <button
                style="background-color: white;"
                type="button"
              >
                Dinner
              </button>
              <button
                style="background-color: white;"
                type="button"
              >
                Lunch
              </button>
              <button
                style="background-color: white;"
                type="button"
              >
                Breakfast
              </button>
              <button
                style="background-color: white;"
                type="button"
              >
                Snack
              </button>
              <button
                style="background-color: white;"
                type="button"
              >
                Dessert
              </button>
            </div>
          </div>
          <label
            class="recipe-description"
          >
            Description :
             
            <textarea
              autocomplete="off"
              cols="50"
              name="description"
              rows="4"
            />
          </label>
        </div>
        <div>
          <label
            class="content-portions"
          >
            Number of portions :
             
            <select
              name="portions"
            >
              <option
                value="1"
              >
                1
              </option>
              <option
                value="2"
              >
                2
              </option>
              <option
                value="3"
              >
                3
              </option>
              <option
                selected=""
                value="4"
              >
                4
              </option>
              <option
                value="5"
              >
                5
              </option>
              <option
                value="6"
              >
                6
              </option>
              <option
                value="7"
              >
                7
              </option>
              <option
                value="8"
              >
                8
              </option>
              <option
                value="9"
              >
                9
              </option>
              <option
                value="10"
              >
                10
              </option>
            </select>
          </label>
          <div
            class="content-ingredients"
          >
            <table
              class="table-ingredients"
            >
              <tbody>
                <tr
                  class="table-headerRow"
                >
                  <th
                    colspan="3"
                  >
                    Ingredients
                  </th>
                  <th>
                    <button
                      type="button"
                    >
                      + Add
                    </button>
                  </th>
                </tr>
                <tr
                  class="table-headerRow"
                >
                  <th>
                    Amount
                  </th>
                  <th>
                    Unit
                  </th>
                  <th>
                    Name
                  </th>
                  <th />
                </tr>
                <tr
                  class="table-contentRow"
                >
                  <td>
                    <div>
                      <input
                        autocomplete="off"
                        min="0"
                        name="amount"
                        step="any"
                        type="number"
                        value="0"
                      />
                    </div>
                  </td>
                  <td>
                    <div>
                      <select
                        autocomplete="off"
                        name="unit"
                      >
                        <option
                          value=""
                        >
                          --option--
                        </option>
                        <option
                          value="cup"
                        >
                          cup
                        </option>
                        <option
                          value="package"
                        >
                          package
                        </option>
                      </select>
                    </div>
                  </td>
                  <td>
                    <div>
                      <input
                        autocomplete="off"
                        name="ingredient"
                        required=""
                        type="text"
                        value=""
                      />
                    </div>
                  </td>
                  <td>
                    <button
                      type="button"
                    >
                      Delete
                    </button>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
        <div
          class="btn-group"
        >
          <button
            style="opacity: 1;"
            type="submit"
          >
            Save
          </button>
          <button
            class="cancelButton"
            type="button"
          >
            Cancel
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
